hw/dv/tools/rules.mk

```diff
@@ -44,24 +44,22 @@ pre_run:
 	/bin/bash ${MAKE_ROOT}/run_dir_limiter ${RUN_PATH} ${RUN_DIR_LIMIT}
 	env > ${RUN_DIR}/env_vars
 
-# TODO: The rm -rf SW_BUILD_DIR below is necessary because currently
-#       sw build dependency does not appear to be fully working with the DV flow
-#       This should be investigated later or replaced with the new build system
 sw_build: pre_run
 ifneq (${SW_NAME},)
-	rm -rf ${SW_BUILD_DIR}
-	mkdir -p ${SW_BUILD_DIR}
-	$(MAKE) -C $(SW_ROOT_DIR)/device \
-	  TARGET=dv \
-	  SW_DIR=boot_rom \
-	  SW_BUILD_DIR=$(SW_BUILD_DIR)/rom \
-	  MAKEFLAGS="$(SW_OPTS)"
-	$(MAKE) -C $(SW_ROOT_DIR)/device \
-	  TARGET=dv \
-	  SW_DIR=$(SW_DIR) \
-	  SW_NAME=$(SW_NAME) \
-	  SW_BUILD_DIR=$(SW_BUILD_DIR)/sw \
-	  MAKEFLAGS="$(SW_OPTS)"
+	# NOTE: Pass -f, since we're going to be re-building everything every time,
```

Comment:
@mcy 
in theory, the way this works for DV is that we will fire off a huge batch of jobs in parallel.
So a particular test or C program, may theoretically be compiled with different seed options that are passed in.

It is also possible that instead of compiling a bunch of different versions, we use the DV backend to pass in values through back-channel communication, so that could work too.  

It seems like meson basically compiles all of our test cases at once, so we'll probably want to tweak our DV flow a bit to do that first, and then start running tests eventually.

Reply:
Well, right now the way I think it works is:
- You fire a ton of parallel make executions.
- Each one configures its copy of Meson with whatever flags it wants.
- Each builds everything.
- The request targets are copied out.

This is fine right now, because all of the configured tests in Make don't rely on passing flags to the C compiler. Meson doesn't have a good way of doing this, so what will eventually happen (in a followup, because the current setup is "good enough" and getting rid of Make is becoming a big blocker for SW):
- DV Make issues a ninja invocation for exactly one SW target.
- Passing a particular set of C compiler flags to a particular binary will be a separate SW target, i.e.,
```
executable(
  my_dv_test_with_cool_flags,
  sources: [my_dv_test.c],
  c_flags: ['my_cool_flags_for_this_test'],
)
```
This is necessary so that Meson actually knows about these flags, rather than having them transmitted over a back-channel.

Whether each ninja invocation uses its own copy of the Meson build tree is... unclear. Since none of the DV settings will be global, this would help make the build process much much faster (though I suspect that it's negligible, since the actual test execution is the largest time sink...).

Reply:
Well, right now the way I think it works is:
- You fire a ton of parallel make executions.
- Each one configures its copy of Meson with whatever flags it wants.
- Each builds everything.
- The request targets are copied out.

This is fine right now, because all of the configured tests in Make don't rely on passing flags to the C compiler. Meson doesn't have a good way of doing this, so what will eventually happen (in a followup, because the current setup is "good enough" and getting rid of Make is becoming a big blocker for SW):
- DV Make issues a ninja invocation for exactly one SW target.
- Passing a particular set of C compiler flags to a particular binary will be a separate SW target, i.e.,
```
executable(
  my_dv_test_with_cool_flags,
  sources: [my_dv_test.c],
  c_flags: ['my_cool_flags_for_this_test'],
)
```
This is necessary so that Meson actually knows about these flags, rather than having them transmitted over a back-channel.

Whether each ninja invocation uses its own copy of the Meson build tree is... unclear. Since none of the DV settings will be global, this would help make the build process much much faster (though I suspect that it's negligible, since the actual test execution is the largest time sink...).